libminifi/include/utils/StringUtils.h

```diff
@@ -187,48 +187,40 @@ constexpr size_t size(const std::basic_string_view<CharT>& str) noexcept { retur
 
 namespace detail {
 // implementation detail of join_pack
-template<typename CharT>
-struct str_detector {
-  template<typename Str>
-  using valid_string_t = decltype(std::declval<std::basic_string<CharT>>().append(std::declval<Str>()));
+template<typename Str, typename CharT>
+concept valid_string = requires(std::basic_string<CharT> left, Str appended) {
+    left.append(appended);
 };
 
-template<typename ResultT, typename CharT, typename... Strs>
-using valid_string_pack_t = std::enable_if_t<(meta::is_detected_v<str_detector<CharT>::template valid_string_t, Strs> && ...), ResultT>;
+// normally CharT would be at the end, because we're checking that Strs are valid strings for CharT,
+// but the syntax needs the parameter pack at the end
+template<typename CharT, typename... Strs>
+concept valid_string_pack = (valid_string<Strs, CharT> && ...);
 
-template<typename CharT, typename... Strs, valid_string_pack_t<void, CharT, Strs...>* = nullptr>
-static std::basic_string<CharT> join_pack(const Strs&... strs) {
-  std::basic_string<CharT> result;
-  result.reserve((size(strs) + ...));
-  (result.append(strs), ...);
-  return result;
-}
+template<typename Str> struct char_type_of_impl {};
+template<typename CharT> struct char_type_of_impl<std::basic_string<CharT>> : std::type_identity<CharT> {};
+template<typename CharT> struct char_type_of_impl<CharT*> : std::type_identity<std::remove_cv_t<CharT>> {};
+template<typename CharT> struct char_type_of_impl<std::basic_string_view<CharT>> : std::type_identity<CharT> {};
+template<typename Str> using char_type_of = typename char_type_of_impl<std::decay_t<Str>>::type;
```

Comment:
Decay here is intentional: not only do we want to ignore cv-qualifiers and references, but `char` arrays should be decayed to pointers, to match the second specialization above.